File: palindromePermutationII/palindromePermutationII.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<string> generatePalindromes(string s) {
        vector<string> res;
        unordered_map<char, int> m;
        for (char c : s) m[c]++;
        string half;
        string mid = "";
        int odd = 0;
        for (auto& p : m) {
            if (p.second % 2) {
                mid += p.first;
                odd++;
            }
            half += string(p.second / 2, p.first);
            if (odd > 1) return res;
        }
        vector<bool> visited(half.length(), false);
        string path;
        generatePerm(res, mid, half, visited, path);
        return res;
    }
    void generatePerm(vector<string>& res, string& mid, string& half, vector<bool>& visited, string& path) {
        if (path.length() == half.length()) {
            string r_path = path;
            reverse(r_path.begin(), r_path.end());
            res.push_back(path + mid + r_path);
            return;
        }
        for (int i = 0; i < half.length(); i++) {
            if (i != 0 && half[i] == half[i - 1] && !visited[i - 1])
                continue;
            if (!visited[i]) {
                path += half[i];
                visited[i] = true;
                generatePerm(res, mid, half, visited, path);
                path.pop_back();
                visited[i] = false;
            }
        }
    }
};

int main() {
    Solution s;
    string str = "aaabb";
    for (string ss : s.generatePalindromes(str))
        cout << ss << " ";
    cout << endl;
    return 0;
}

File: 267.palindromePermutationII/PalindromePermutationII.java
import java.util.*;

class PalindromePermutationII {
    public List<String> generatePalindromes(String s) {
        List<String> res = new ArrayList<>();
        Map<Character, Integer> m = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            m.put(c, m.containsKey(c) ? m.get(c) + 1 : 1);
        }
        String mid = "";
        int odd = 0;
        List<Character> half = new ArrayList<>();
        for (Map.Entry<Character, Integer> entry : m.entrySet()) {
            char key = entry.getKey();
            int val = entry.getValue();
            if (val % 2 != 0) {
                mid += key;
                odd++;
            }
            for (int i = 0; i < val / 2; i++)
                half.add(key);
            if (odd > 1)
                return res;
        }
        boolean[] visited = new boolean[half.size()];
        StringBuilder path = new StringBuilder();
        genPerm(half, mid, visited, path, res);
        return res;
    }
    void genPerm(List<Character> half, String mid, boolean[] visited, StringBuilder path, List<String> res) {
        if (path.length() == half.size()) {
            res.add(path.toString() + mid + path.reverse().toString());
            path.reverse();
            return;
        }
        for (int i = 0; i < half.size(); i++) {
            if (i != 0 && half.get(i).equals(half.get(i - 1)) && !visited[i - 1])
                continue;
            if (!visited[i]) {
                path.append(half.get(i));
                visited[i] = true;
                genPerm(half, mid, visited, path, res);
                path.deleteCharAt(path.length() - 1);
                visited[i] = false;
            }
        }
    }
	public static void main(String[] args) {
		PalindromePermutationII s = new PalindromePermutationII();
		String str = "aaabb";
		for (String ss : s.generatePalindromes(str))
			System.out.print(ss + " ");
		System.out.println();
	}
}
